README.md
# SharedNearestNeighbors
A Shared Nearest Neighbors (Jarvis-patrick) clustering implementation. This code is basically a wrapper of sklearn DBSCAN, implementing the neighborhood similarity as a metric.

## Examples

The implementation follows the syntax of scikit-learn clustering classes.
See example notebook for more information.

```python
from SNN import SharedNearestNeighbor as SNN
```

```python
X = np.random.rand(100,2)
snn = SNN(n_neighbors = 40,eps = 32, min_samples = 10)
y_pred = snn.fit_predict(X)
```

![](imgs/D31.png)


```python
snn = SNN(n_neighbors = 35,eps = 25, min_samples = 8)
y_pred = snn.fit_predict(X)

```
![](imgs/aggregation.png)


```python
snn = SNN(n_neighbors = 160,eps = 120, min_samples = 30)
y_pred = snn.fit_predict(X)

```
![](imgs/t4-8k.png)


```python
snn = SNN(n_neighbors = 160,eps = 120, min_samples = 35)
y_pred = snn.fit_predict(X)

```
![](imgs/t7-10k.png)


```python
snn = SNN(n_neighbors = 4, eps = 2, min_samples = 2)
y_pred = snn.fit_predict(X)
```
![](imgs/3-spiral.png)

```python
snn = SNN(n_neighbors = 15, eps = 7, min_samples = 12)
y_pred = snn.fit_predict(X)
```
![](imgs/DS-850.png)

## To-do list

- [ ] Add weighted option for similarity between neighbors


## References

Jarvis, R. A., & Patrick, E. A. (1973). Clustering using a similarity measure based on shared near neighbors. IEEE Transactions on computers, 100(11), 1025-1034.

Ertöz, Levent, Michael Steinbach, and Vipin Kumar. "Finding clusters of different sizes, shapes, and densities in noisy, high dimensional data." Proceedings of the 2003 SIAM international conference on data mining. Society for Industrial and Applied Mathematics, 2003.


HISTORY.rst
=======
History
=======

0.1.1 (2022-01-09)
------------------

* Added docs and CI pipeline



0.1.0 (2022-01-06)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for shared_nearest_neighbors."""


tests/test_shared_nearest_neighbors.py
#!/usr/bin/env python
"""Tests for `shared_nearest_neighbors` package."""

import pytest
import numpy as np

from shared_nearest_neighbors.shared_nearest_neighbors import SNN
from sklearn.cluster.tests.common import generate_clustered_data


@pytest.fixture
def n_clusters():
    return 3


@pytest.fixture
def clustered_data(n_clusters):
    return generate_clustered_data(n_clusters=n_clusters)


@pytest.fixture
def one_dimensional_data():
    return np.array(
        [
            [1],
            [3],
            [6],
        ]
    )


def test_bad_eps():

    with pytest.raises(ValueError):
        snn = SNN(n_neighbors=10, eps=10, min_samples=30, metric="euclidean")

def test_snn_clustering(clustered_data, n_clusters):
    snn = SNN(n_neighbors=10, eps=5, min_samples=10, metric="euclidean")
    snn.fit(clustered_data)
    
    labels = snn.labels_
    n_clusters_out = len(set(labels)) - int(-1 in labels)
    assert n_clusters == n_clusters_out


def test_similarity_matrix(one_dimensional_data):
    n_neighbors = 2
    snn = SNN(n_neighbors=n_neighbors, eps=0.5, min_samples=1)
    dissimilarity_matrix = snn.neighborhood_dissimilarity_matrix(one_dimensional_data)
    
    shared_neighbors = np.array([[2, 2, 1], [2, 2, 1], [1, 1, 2]])
    expected = n_neighbors - shared_neighbors

    assert np.all(np.isclose(dissimilarity_matrix.toarray(), expected))


SNN.py
from sklearn.cluster import DBSCAN
from sklearn.neighbors import NearestNeighbors
import numpy as np


class SharedNearestNeighbor:
    """
    Shared Nearest Neighbor clustering  algorithm for finding clusters or different sizes, shapes and densities in
    noisy, high-dimensional datasets. 

    The algorithm can be seen as a variation of DBSCAN which uses neighborhood similairty as a metric.
    It does not have a hard time detecting clusters of different densities as DBSCAN, and keeps its advantages.
    
    References:

    Ertöz, L., Steinbach, M., & Kumar, V. (2003, May). Finding clusters of different sizes, shapes, and densities in noisy, high dimensional data. In Proceedings of the 2003 SIAM international conference on data mining (pp. 47-58). Society for Industrial and Applied Mathematics.

    Ertoz, Levent, Michael Steinbach, and Vipin Kumar. "A new shared nearest neighbor clustering algorithm and its applications." Workshop on clustering high dimensional data and its applications at 2nd SIAM international conference on data mining. 2002.

    """

    def __init__(self,n_neighbors = 7, eps = 5, min_samples = 5, weighted = False, n_jobs = 1, algorithm = 'auto', leaf_size = 30, metric = 'minkowski', p = 2, metric_params = None):
        """
        @param n_neighbors: number of neighbors to consider when calculating the shared nearest neighbors
        @param eps: threshold on the number of neighbors
        @param min_samples: minimum number  of samples that share at least eps neighbors so that a point can be considered  a core point
        @param weighted: if True, uses weighed version o Shared Nearest Neighbors, and eps must be between 0 and 1.
        @param n_jobs: number of parallel jobs
        @param algorithm: parameter for Nearest Neighbors calculation, please see scikit-learn documentation
        @param leaf_size: parameter for Nearest Neighbors calculation, please see scikit-learn documentation
        @param metric: parameter for Nearest Neighbors, please see scikit-learn documentation
        @param p: param for Nearest Neighbors, please see scikit-learn documentation
        @param metric_params: param for Nearest Neighbors, please see scikit-learn documentation
        @return an instance of SharedNearestNeighbor class
        """
        if weighted and eps >= 1:
            raise ValueError("For weighed SNN, please define a eps value between 0 and 1.")
        if eps < 0:
            raise ValueError("Eps must be positive.")

        self.neigh =  NearestNeighbors(n_neighbors = n_neighbors, n_jobs = n_jobs, algorithm= algorithm, leaf_size= leaf_size, metric= metric, p = p, metric_params= metric_params)
        self.eps = eps
        self.min_samples = min_samples
        self.n_jobs = n_jobs
        self.n_neighbors = n_neighbors
        self.labels_ = []
        self.components_ = []
        self.core_sample_indices_ = []
        self.weighted = weighted

        

    def fit(self,X):
        """
        @param X: array of shape (n_samples, n_features)
        @return self
        """
        if self.weighted:
            graph, max_similarity = self._inner_weighted(X)    
        else: 
            graph, max_similarity = self._inner(X)
        

        self.dbscan = DBSCAN(eps = max_similarity - self.eps, min_samples = self.min_samples,metric = "precomputed", n_jobs = self.n_jobs)
        self.dbscan.fit(self.similarity_matrix)
        self.labels_ = self.dbscan.labels_
        self.components_ = self.dbscan.components_
        self.core_sample_indices_ = self.dbscan.core_sample_indices_
        return self

    def _inner(self,X): 
        """
        Calculates the similarity matrix of the dataset

        @param X: input data matrix of shape (n_samples,n_features)
        @return graph, an sparse similarity distance matrix, and the value of the maximum similarity

        """
        self.neigh.fit(X)
        graph = self.neigh.kneighbors_graph(X)
        self.similarity_matrix = graph*graph.transpose()
        self.mask = self.similarity_matrix > self.eps
        max_similarity = self.similarity_matrix[0,0]
        self.similarity_matrix.data = max_similarity - self.similarity_matrix.data
        return graph, max_similarity

    def _inner_weighted(self,X):
        """
        Calculates the similarity matrix of the dataset based on weighed edges

        @param X: input data matrix of shape (n_samples,n_features)
        @return graph, an sparse similarity distance matrix, and the value of the maximum similarity
        """
        self.neigh.fit(X)
        graph = self.neigh.kneighbors_graph(X, mode = "distance")
        graph.data = np.reshape(self.n_neighbors - np.argsort(np.argsort(graph.data.reshape((-1,self.n_neighbors)))),(-1,))
        self.similarity_matrix = graph*graph.transpose()
        self.mask = self.similarity_matrix > self.eps
        max_similarity = self.similarity_matrix[0,0]
        assert(max_similarity == np.max(self.similarity_matrix))
        self.similarity_matrix.data = max_similarity - self.similarity_matrix.data
        self.similarity_matrix.data = self.similarity_matrix.data/max(self.similarity_matrix.data)
        print("Biggest distance: ", np.max(self.similarity_matrix.data))
        print("Mean distance:",np.mean(self.similarity_matrix.data))
        print("Median distance:",np.median(self.similarity_matrix.data))
        max_similarity = 1
        return graph, max_similarity

    def fit_predict(self,X):
        """
        @param X: array of shape (n_samples, n_features)
        @return y: array of shape (n_samples), containing the predicted cluster labels
        """
        if self.weighted:
            graph, max_similarity = self._inner_weighted(X)    
        else: 
            graph, max_similarity = self._inner(X)

        self.dbscan = DBSCAN(eps = max_similarity - self.eps, min_samples = self.min_samples,metric = "precomputed", n_jobs = self.n_jobs)
        y = self.dbscan.fit_predict(self.similarity_matrix)
        self.labels_ = self.dbscan.labels_
        self.components_ = self.dbscan.components_
        self.core_sample_indices_ = self.dbscan.core_sample_indices_
        return y
        
        


shared_nearest_neighbors/shared_nearest_neighbors.py
import numpy as np
from sklearn.base import ClusterMixin, BaseEstimator
from sklearn.neighbors import NearestNeighbors
from sklearn.cluster import DBSCAN
from scipy.sparse import csr_matrix, spdiags

def snn_dissimilarity_func(graph : csr_matrix, n_neighbors : int, *args, **kwargs) -> csr_matrix:
    """Default SNN dissimilarity function

    Computes the dissimilarity between two points in terms of shared nearest neighbors

    Args:
        graph (scipy.sparse.csr_matrix): sparse matrix with dimensions (n_samples, n_samples),
         where the element ij represents the distance between the point i and j 
        n_neighbors (int): number of neighbors in the k-neighborhood search
    """ 

    graph.data[graph.data > 0] = 1
    n_samples = graph.shape[0]

    # Add the point as its own neighbor
    graph += spdiags(np.ones(n_samples), diags=0, m=n_samples, n=n_samples)
    matrix = graph * graph.transpose()
    matrix.sort_indices()

    # The lower the "closer"
    matrix.data = n_neighbors - matrix.data

    return matrix


class SNN(ClusterMixin, BaseEstimator):
    def __init__(
        self,
        n_neighbors=7,
        eps=5,
        min_samples=5,
        algorithm="auto",
        leaf_size=30,
        metric="euclidean",
        p=None,
        metric_params=None,
        dissimilarity_func=snn_dissimilarity_func,
        n_jobs=None,
    ):
        """Shared Nearest Neighbor clustering  algorithm for finding clusters or different sizes, shapes and densities in
        noisy, high-dimensional datasets.


        The algorithm can be seen as a variation of DBSCAN which uses neighborhood similairty as a metric.
        It does not have a hard time detecting clusters of different densities as DBSCAN, and keeps its advantages.


        Parameters
        ----------
        n_neighbors : int, optional
            The number of neighbors to construct the neighborhood graph, including the point itself. By default 7

        eps : int, optional
            The minimum number of neighbors two points have to share in order to be
            connected by an edge in the neighborhood graph. This value has to be smaller
            than n_neighbors. By default 5

        min_samples : int, optional
            The number of samples (or total weight) in a neighborhood for a point
            to be considered as a core point. This includes the point itself, by default 5

        algorithm : {'auto', 'ball_tree', 'kd_tree', 'brute'}, optional
            The algorithm to be used by the NearestNeighbors module
            to compute pointwise distances and find nearest neighbors.
            See NearestNeighbors module documentation for details., by default "auto"

        leaf_size : int, optional
            [description], by default 30

        metric : str, or callable
            The metric to use when calculating distance between instances in a
            feature array. If metric is a string or callable, it must be one of
            the options allowed by :func:`sklearn.metrics.pairwise_distances` for
            its metric parameter.
            If metric is "precomputed", X is assumed to be a distance matrix and
            must be square. X may be a :term:`Glossary <sparse graph>`, in which
            case only "nonzero" elements may be considered neighbors for DBSCAN.
            Default to "euclidean"

        p : int, optional
            The power of the Minkowski metric to be used to calculate distance
            between points. If None, then ``p=2`` (equivalent to the Euclidean
            distance).

        metric_params : [type], optional
            Additional keyword arguments for the metric function., by default None

        n_jobs : int, optional
            The number of parallel jobs to run.
            ``None`` means 1 unless in a :obj:`joblib.parallel_backend` context.
            ``-1`` means using all processors. See :term:`Glossary <n_jobs>`
            for more details. Default None.

        dissimilarity_func: Callable, optional
            A function that receives two inputs: scipy.sparse.csr_matrix with the k-neighbors distance and the n_neighbors attribute;
            and returns another csr_matrix

        Attributes
        ----------
        
        neigh : sklearn.neighbors.NearestNeighbors 
        
        dbscan : sklearn.cluster.DBSCAN

        labels_ : ndarray of shape (n_samples)
            Cluster labels for each point in the dataset given to fit().
            Noisy samples are given the label -1.
            
        components_ : ndarray of shape (n_core_samples, n_features)

        Copy of each core sample found by training.

        core_samples_indices_ : ndarray of shape (n_core_samples,)
            Indices of core samples.

        dissimilarity_matrix : scipy.sparse.csr_matrix 
            containing the dissimilarity between points

        References
        ----------

        Ertöz, L., Steinbach, M., & Kumar, V. (2003, May). Finding clusters of different sizes, shapes, and densities in noisy, high dimensional data. In Proceedings of the 2003 SIAM international conference on data mining (pp. 47-58). Society for Industrial and Applied Mathematics.
        Ertoz, Levent, Michael Steinbach, and Vipin Kumar. "A new shared nearest neighbor clustering algorithm and its applications." Workshop on clustering high dimensional data and its applications at 2nd SIAM international conference on data mining. 2002.


        """

        if eps <= 0:
            raise ValueError("Eps must be positive.")
        if eps >= n_neighbors and dissimilarity_func == snn_dissimilarity_func:
            raise  ValueError("Eps must be smaller than n_neighbors.")

        self.eps = eps
        self.min_samples = min_samples
        self.n_jobs = n_jobs
        self.n_neighbors = n_neighbors
        self.algorithm = algorithm
        self.leaf_size = leaf_size
        self.metric = metric
        self.p = p
        self.metric_params = metric_params
        self.dissimilarity_func = dissimilarity_func
        self.neigh = NearestNeighbors(
            n_neighbors=self.n_neighbors,
            n_jobs=self.n_jobs,
            algorithm=self.algorithm,
            leaf_size=self.leaf_size,
            metric=self.metric,
            p=self.p,
            metric_params=self.metric_params,
        )

        # Reasoning behind eps=self.n_neighbors - self.eps:
        # In DBSCAN, eps is an upper bound of the distance between two points.
        # In terms of similarity, it would an "lower bound" on the similarity
        # or, once again, upper bound on the difference between the max similarity value and
        # the similarity between two points
        self.dbscan = DBSCAN(
            eps=self.n_neighbors - self.eps,
            min_samples=self.min_samples,
            metric="precomputed",
            n_jobs=self.n_jobs,
        )

    def fit(self, X, y=None, sample_weight=None):
        """Perform SNN clustering from features or distance matrix

        First calls NearestNeighbors to construct the neighborhood graph considering the params
        n_neighbors, n_jobs, algorithm, leaf_size, metric, p, metric_params

        Parameters
        ----------
        X : {array-like, sparse matrix} of shape (n_samples, n_features), or (n_samples, n_samples)
            Training instances to cluster, or distances between instances if
            ``metric='precomputed'``. If a sparse matrix is provided, it will
            be converted into a sparse ``csr_matrix``.

        y : Ignored
            Not used, present here for API consistency by convention.
        sample_weight : array-like of shape (n_samples,), default=None
            Weight of each sample, such that a sample with a weight of at least
            ``min_samples`` is by itself a core sample; a sample with a
            negative weight may inhibit its eps-neighbor from being core.
            Note that weights are absolute, and default to 1.

        Returns
        -------
        self : object
            Returns a fitted instance of self.
        """

        self.dissimilarity_matrix = self.neighborhood_dissimilarity_matrix(X)

        self.dbscan.fit(self.dissimilarity_matrix, sample_weight=sample_weight)

        return self

    @property
    def labels_(self):
        return self.dbscan.labels_
    
    @property
    def components_(self):
        return self.dbscan.components_

    @property
    def core_sample_indices_(self):
        return self.dbscan.core_sample_indices_


    def neighborhood_dissimilarity_matrix(self, X) -> csr_matrix:
        """Neighborhood similarity matrix

        Computes the sparse neighborhood similarity matrix

        Parameters
        ----------
        X : {array-like, sparse matrix} of shape (n_samples, n_features), or (n_samples, n_samples)
            Training instances to cluster, or distances between instances if
            ``metric='precomputed'``. If a sparse matrix is provided, it will
            be converted into a sparse ``csr_matrix``.

        Returns
        -------
        csr_matrix
            Sparse matrix of shape (n_samples, n_samples)
        """

        self.neigh.fit(X)
        graph = self.neigh.kneighbors_graph(X, mode="distance")
        dissimilarity_matrix = self.dissimilarity_func(graph, self.n_neighbors)
        return dissimilarity_matrix